docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+- Rewriting an effect and/or performing actions with the effect value and then performing the originally intended effect. This technique is conceptually similar to the middleware pattern commonly used in web applications.
+
+Let's explore both of these effect interpretation strategies with a small motivating example:
+
+✨ We would like to implement a client library for an [HTTP-based API that provides interesting cat facts](https://alexwohlbruck.github.io/cat-facts/docs/). ✨
+
+Let's break down some of the properties of the API client that would be desirable for a production use case:
+
+1. We would like to have our cat facts API be able to support different cat fact data sources in the future.
+2. We would like to be able to mock failure conditions (such as network connectivity issues) for testing purposes.
+3. We would like to be able to track timing metrics for how quickly we can retrieve cat facts.
+
+### Initial setup
+
+``` haskell
+{-# LANGUAGE ExistentialQuantification, DeriveAnyClass, DeriveFunctor,
+DeriveGeneric, DerivingStrategies, FlexibleContexts, FlexibleInstances,
+GeneralizedNewtypeDeriving, OverloadedStrings, MultiParamTypeClasses,
+RankNTypes, TypeApplications, TypeOperators, UndecidableInstances #-}
+module CatFacts
+    ( main
+    ) where
+-- from base
+import Control.Applicative
+import Control.Exception (throwIO)
+import GHC.Generics (Generic1)
+-- from fused-effects
+import Control.Algebra
+import Control.Carrier.Reader
+import Control.Carrier.Error.Either
+import Control.Carrier.Interpret
+-- from transformers
+import Control.Monad.IO.Class
+-- From aeson
+import Data.Aeson
+-- From bytestring
+import qualified Data.ByteString.Char8 as B
+import qualified Data.ByteString.Lazy as L
+-- From time
+import Data.Time.Clock
+-- From http-client
+import qualified Network.HTTP.Client as HTTP
+import Network.HTTP.Client.Internal (Response(..), ResponseClose(..))
+-- From http-client-tls
+import qualified Network.HTTP.Client.TLS as HTTP
+-- From http-status
+import Network.HTTP.Types.Header
+import Network.HTTP.Types.Status
+import Network.HTTP.Types.Version
+```
+
+Since one of the best parts about effects is being able to think at a domain language level,
+let's start with defining the desired data which we wish to retrieve and an interface that
+feels natural to work with:
+
+``` haskell
+-- | The basic fact that we will retrieve.
+newtype CatFact = CatFact
+  { catFact :: String
+  } deriving (Show)
+
+instance FromJSON CatFact where
+  parseJSON = withObject "CatFact" (\o -> CatFact <$> o .: "text")
+
+-- | Our high level effect type that will be able to target different data sources.
+data CatFactClient m k
+  = ListFacts Int {- ^ Number of facts to fetch -} ([CatFact] -> m k)
+  deriving stock (Functor, Generic1)
+  deriving anyclass (HFunctor, Effect)
```

Comment:
1. `HFunctor` is gone as of #295.

2. I’d strongly prefer to avoid `DeriveAnyClass` & `DerivingStrategies` since I think they present an unnecessarily complicated picture of the use of the library; `instance Effect` will suffice.

(Ditto for the other effects you’ve defined, naturally.)